src/Symfony/Component/Config/Definition/EnumNode.php

```diff
@@ -78,13 +104,48 @@ protected function finalizeValue(mixed $value): mixed
     {
         $value = parent::finalizeValue($value);
 
-        if (!\in_array($value, $this->values, true)) {
-            $ex = new InvalidConfigurationException(\sprintf('The value %s is not allowed for path "%s". Permissible values: %s', json_encode($value), $this->getPath(), $this->getPermissibleValues(', ')));
-            $ex->setPath($this->getPath());
+        if ($this->enumFqcn) {
+            if (is_a($this->enumFqcn, \BackedEnum::class, true)) {
+                if (\is_string($value) || \is_int($value)) {
+                    try {
+                        $case = $this->enumFqcn::tryFrom($value);
+                    } catch (\TypeError) {
+                        throw new InvalidConfigurationException(\sprintf('The value could not be casted to a case of the "%s" enum. Is the value the same type as the backing type of the enum?', $this->enumFqcn));
+                    }
+
+                    if (null !== $case) {
+                        return $case;
+                    }
+                } elseif ($value instanceof \UnitEnum && !$value instanceof $this->enumFqcn) {
+                    throw new InvalidConfigurationException(\sprintf('The value should be part of the "%s" enum, got a value from the "%s" enum.', $this->enumFqcn, get_debug_type($value)));
+                }
+            }
+
+            if ($value instanceof $this->enumFqcn) {
+                return $value;
+            }
 
-            throw $ex;
+            throw $this->createInvalidValueException($value);
+        }
+
+        if (!\in_array($value, $this->values, true)) {
+            throw $this->createInvalidValueException($value);
         }
 
         return $value;
     }
+
+    private function createInvalidValueException(mixed $value): InvalidConfigurationException
+    {
+        if ($this->enumFqcn) {
+            $message = \sprintf('The value %s is not allowed for path "%s". Permissible values: %s (cases of the "%s" enum).', json_encode($value), $this->getPath(), $this->getPermissibleValues(', '), $this->enumFqcn);
+        } else {
+            $message = \sprintf('The value %s is not allowed for path "%s". Permissible values: %s.', json_encode($value), $this->getPath(), $this->getPermissibleValues(', '));
+        }
```

Comment:
What if our invalid value is not JSON-encodable?

Reply:
I think we're safe here, because `EnumNode` constructor ensure given values are either null or scalar

Reply:
It's still not safe if a string isn't utf-8.
It'd suggest displaying its type if it's not a string nor an int, and displaying the value otherwise:
- Value of type "foo"
- Value 123
- Value "some string"